KD_Lib/Pruning/lottery_tickets/lottery_tickets.py

```diff
@@ -0,0 +1,204 @@
+                outputs = self.model(data)
+
+                if isinstance(outputs, tuple):
+                    outputs = outputs[0]
+
+                test_loss += self.loss_fn(outputs, targets).item()
+                pred = outputs.argmax(dim=1, keepdim=True)
+                correct += pred.eq(targets.data.view_as(pred)).sum().item()
+
+            test_loss /= len(self.test_loader.dataset)
+            test_acc = 100.0 * correct / len(self.test_loader.dataset)
+
+        return test_loss, test_acc
+
+    def _train_pruned_model(self):
+        eps = 1e-6
+        self.model.train()
+        correct = 0
+
+        step = 0
+        for data, targets in self.train_loader:
+            self.optimizer.zero_grad()
+            data, targets = data.to(self.device), targets.to(self.device)
+            outputs = self.model(data)
+
+            if isinstance(outputs, tuple):
+                outputs = outputs[0]
+
+            train_loss = self.loss_fn(outputs, targets)
+            train_loss.backward()
+
+            pred = outputs.argmax(dim=1, keepdim=True)
+            correct += pred.eq(targets.data.view_as(pred)).sum().item()
+
+            train_acc = 100.0 * correct / len(self.train_loader.dataset)
+
+            # print(f"Training Step: {step} | Loss: {train_loss.item()} | Accuracy: {train_acc}")
+
+            for name, param in self.model.named_parameters():
+                if "weight" in name:
+                    param_data = param.data.cpu().numpy()
+                    param_grad = param.grad.data.cpu().numpy()
+                    param_grad = np.where(param_data < eps, 0, param_grad)
+                    param.grad.data = torch.from_numpy(param_grad).to(self.device)
+
+            self.optimizer.step()
+            step += 1
+
+        train_acc = 100.0 * correct / len(self.train_loader.dataset)
+        return train_loss.item(), train_acc
+
+    def _save_model(self, prune_it, it):
+        file_name = f"{os.getcwd()}/{prune_it}_{it}_pruned_model.pth.tar"
+        torch.save(self.model, file_name)
+
+    def get_pruning_statistics(model_path=None):
+        if model_path is not None:
+            model = torch.load(model_path)
+        else:
+            model = self.model
+        nonzero = total = 0
+        for name, p in model.named_parameters():
+            tensor = p.data.cpu().numpy()
+            nz_count = np.count_nonzero(tensor)
+            total_params = np.prod(tensor.shape)
+            nonzero += nz_count
+            total += total_params
+            print(
+                f"{name:20} | nonzeros = {nz_count:7} / {total_params:7} ({100 * nz_count / total_params:6.2f}%) | total_pruned = {total_params - nz_count :7} | shape = {tensor.shape}"
+            )
+        print(
+            f"alive: {nonzero}, pruned : {total - nonzero}, total: {total}, Compression rate : {total/nonzero:10.2f}x  ({100 * (total-nonzero) / total:6.2f}% pruned)"
+        )
+        return round((nonzero / total) * 100, 1)
```

Comment:
I guess do one thing. If `model_path` is `None` then print statistics for all the iterations 
(Of what I think from this code it is printing only for the one model if it it None)